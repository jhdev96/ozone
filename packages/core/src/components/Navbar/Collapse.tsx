import React from 'react';
import styles from './index.module.scss';


interface CollapseProps {
  /** 
   * An HTML id for the Collapse component. 
   * 
   * This is used to set the target element
   * for the `Toggle` component.
   */
  id: string;
  /**
   * The content rendered inside this
   * component.
   */
  children?: React.ReactNode;
}

export default function Collapse({
  id,
  children
}: CollapseProps): JSX.Element 
{
  return (
    <div id={id} className={styles.navCollapse}>
      {children}
    </div>
  )
}

Collapse.displayName = 'NavbarCollapse';