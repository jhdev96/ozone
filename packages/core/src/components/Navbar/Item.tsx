import React from 'react';
import styles from './index.module.scss';


interface ItemProps {
  /** 
   * Function that is called when an
   * item is clicked.
   */
  onClick?: (name?: string) => void;
  /**
   * A name for the item.
   */ 
  name?: string;
  /** 
   * The content rendered inside this component. 
   */
  children?: React.ReactNode;
}
 
export default function Item({
  onClick = () => {},
  name,
  children
}: ItemProps): JSX.Element 
{
  const handleClick = (e: React.MouseEvent<HTMLLIElement>) => {
    const classes = e.currentTarget.classList;

    switch (e.currentTarget.tagName.toLowerCase()) {
      case 'a':
        if (classes.contains(styles.disabled) || onClick) {
          e.preventDefault();
        }
        break;
      default:
        onClick(name);
    }
  }

  return (
    <li 
      className={styles.navItem} 
      onClick={handleClick}
    >
      {children}
    </li>
  )
}

Item.displayName = 'NavbarItem';