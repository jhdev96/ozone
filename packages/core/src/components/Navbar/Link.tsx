import React from 'react';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


export interface LinkProps {
  /** 
   * Sets the `href` attribute for the anchor element. 
   */ 
  to?: string;
  /** 
   * Visually indicates that the link is active.
   */ 
  isActive?: boolean;
  /** 
   * Visually and functionally indicates that the
   * link is disabled. 
   */
  isDisabled?: boolean;
  /** 
   * The link's content.
   */
  children?: React.ReactNode;
}

export default function Link({
  to,
  isActive,
  isDisabled,
  children
}: LinkProps): JSX.Element 
{
  return (
    <a 
      href={to}
      className={classNames(
        styles.navLink, 
        isActive    && styles.active,
        isDisabled  && styles.disabled
      )} 
    >
      {children}
    </a>
  );
}

Link.displayName = 'NavbarLink';