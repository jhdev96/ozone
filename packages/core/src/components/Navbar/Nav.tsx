import React from 'react';
import styles from './index.module.scss';


interface Props {
  /**
   * The items inside the navbar.
   * These should be `Item` components. 
   */
  children?: React.ReactNode;
};

export default function Nav({
  children
}: Props): JSX.Element 
{
  return (
    <ul className={styles.navContent}>
      {children}
    </ul> 
  )
}

Nav.displayName = 'NavbarNav';