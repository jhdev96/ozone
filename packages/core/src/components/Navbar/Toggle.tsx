import React from 'react';
import styles from './index.module.scss';


interface ToggleProps {
  /** 
   * A label used to identify the toggle
   * button for accessibility purposes.
   */
  label?: string;
  /** 
   * The HTML id of the target element.
   * 
   * The target element should be a
   * `Collapse` component. 
   */ 
  targetId: string;
  /** 
   * The content that will be rendered
   * inside the toggle button.
   */ 
  children?: React.ReactNode;
}

export default function Toggle({
  label,
  targetId,
  children
}: ToggleProps): JSX.Element 
{
  const toggleNav = () => {
    const nav = document.getElementById(targetId);
    
    if (!nav?.classList.contains(styles.show)) {
      nav?.classList.add(styles.show);
    } else {
        nav?.classList.remove(styles.show);
    }
  }

  return (
    <button 
      type="button"
      className={styles.navToggle} 
      aria-label={label}
      onClick={toggleNav}
    >
      {children}
    </button>
  )
}

Toggle.displayName = 'NavbarToggle';