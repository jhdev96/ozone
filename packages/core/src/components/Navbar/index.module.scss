@import '../../scss/root';
@import '../../scss/mixins/navbar/navbar.scss';


$mediaMap: ('sm': 480px, 'md': 768px, 'lg': 992px); // for media queries

.nav { 
  display: flex; 
  height: fit-content;
  padding: 0.15rem;
  font-size: 0.85em;

  svg { vertical-align: text-top; }

  /* Nav Elements */
  &Brand {
    padding: 0 0.8em;
    color: var(--black-core);
    font-size: 1.1em;
    font-weight: 600;
    text-decoration: none;
  }

  &Toggle { 
    display: none; 
    margin: 0 0.6em;
    background: none;
    border: 0 !important;
    outline: 0 !important;
    cursor: pointer;
  }

  &Collapse {
    height: fit-content;
    align-self: end;
  }

  &Content {
    display: flex; 
    align-self: center;
  }

  &Item {
    margin: 0 0.60em;
    height: fit-content;
    line-height: 1.4;
    cursor: pointer !important;
  }

  &Link { text-decoration: none; }
}

// justify
.start { justify-content: flex-start; }
.end { justify-content: flex-end; }
.center { justify-content: center; }
.space-between { justify-content: space-between; }

// Orientation
.horizontal {
  .navContent {
    flex-direction: row;
  }
}

.vertical {
  .navContent {
    flex-direction: column;
  }
}

// Theme
.light {
  .navLink:not(.disabled) {
    color: var(--black-2) !important;
  }
}

.dark {
  .navLink:not(.disabled) { 
    color: var(--gray-3); 
  }
  .navBrand { color: #fff; }
}

// isActive
.active { color: var(--blue-core) !important; }

// isDisabled
.disabled {
  color: var(--gray-4) !important;
  cursor: default;
}

// expand
@each $size, $breakpoint in $mediaMap {
  @media (max-width: $breakpoint) {
    @include makeResponsive($size);
  }
}