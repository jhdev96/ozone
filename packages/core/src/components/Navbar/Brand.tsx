import React from 'react';
import { LinkProps } from './Link';
import styles from './index.module.scss';


export default function Brand({
  to,
  children
}: LinkProps): JSX.Element 
{
  return (
    <a 
      className={styles.navBrand} 
      href={to}
    >
      {children}
    </a>
  )
}

Brand.displayName = 'NavbarBrand';