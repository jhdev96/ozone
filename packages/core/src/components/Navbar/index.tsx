import React from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import Item from './Item';
import Link from './Link';
import Nav from './Nav';
import Brand from './Brand';
import Toggle from './Toggle';
import Collapse from './Collapse';
import styles from './index.module.scss';


interface NavbarProps {
  /** 
   * Determines if links are displayed vertically or horizontally. 
   */ 
  orientation?: "horizontal" | "vertical";
  /** 
   * Positions the navbar brand and the items. 
   */ 
  justify?: "start" | "end" | "center" | "space-between";
  /** 
   * Gives the navbar a light or dark theme. 
   */ 
  theme?: ThemeType;
  /** 
   * Determines what screen size should trigger responsiveness. 
   * This should be used along with a Navbar.Collapse component.
   */ 
  expand?: "sm" | "md" | "lg";
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * The content displayed inside the navbar. 
   */
  children?: React.ReactNode;
};

export default function Navbar({
  orientation = "horizontal",
  justify     = "start",
  theme       = "light",
  expand      = "lg",
  dataTestId,
  children
}: NavbarProps): JSX.Element 
{
  return (
    <nav 
      className={classNames(
        styles.nav,
        styles[orientation],
        styles[justify],
        styles[theme],
        styles[expand]
      )}
      data-testid={dataTestId}
    >
      {children} 
    </nav>
  );
}

Navbar.Item = Item;
Navbar.Link = Link;
Navbar.Nav = Nav;
Navbar.Brand = Brand;
Navbar.Collapse = Collapse;
Navbar.Toggle = Toggle;

Navbar.displayName = 'Navbar';