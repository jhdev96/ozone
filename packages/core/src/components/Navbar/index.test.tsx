import React from "react";
import { render } from '@testing-library/react';
import * as icons from 'icons';
import Navbar from ".";


describe('<Navbar />', () => {
  it('renders a navbar with 3 links', () => {
    render(
      <Navbar dataTestId="test">
        <Navbar.Nav>
          <Navbar.Item>
            <Navbar.Link to="#">Link 1</Navbar.Link>
          </Navbar.Item>
          <Navbar.Item>
            <Navbar.Link to="#">Link 2</Navbar.Link>
          </Navbar.Item>
          <Navbar.Item>
            <Navbar.Link to="#">Link 3</Navbar.Link>
          </Navbar.Item>
        </Navbar.Nav>
      </Navbar>
    );

    expect(document.querySelectorAll('.navLink'))
      .toHaveLength(3);
  });

  it('renders a navbar with horizontal orientation by default', () => {
    const {container} = render(
      <Navbar dataTestId="test">
        <Navbar.Nav>
          <Navbar.Item>
            <Navbar.Link to="#">Link 1</Navbar.Link>
          </Navbar.Item>
        </Navbar.Nav>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });

  it('sets the `justify` prop to `start` by default', () => {
    const {container} = render(
      <Navbar dataTestId="test" />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a navbar with a brand link', () => {
    const {container} = render(
      <Navbar>
        <Navbar.Brand>Brand</Navbar.Brand>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });

  it('centers the navbar items', () => {
    const {container} = render(
      <Navbar justify="center">
        <Navbar.Brand>Brand</Navbar.Brand>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });

  it('places the navbar items at the end', () => {
    const {container} = render(
      <Navbar justify="end">
        <Navbar.Brand>Brand</Navbar.Brand>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });

  it('adds space between the brand and the links', () => {
    const {container} = render(
      <Navbar justify="space-between">
        <Navbar.Brand>Brand</Navbar.Brand>
        <Navbar.Nav>
          <Navbar.Item>Item 1</Navbar.Item>
          <Navbar.Item>Item 2</Navbar.Item>
          <Navbar.Item>Item 3</Navbar.Item>
        </Navbar.Nav>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });

  it('sets the `orientation` prop to `vertical`', () => {
    const {container} = render(
      <Navbar orientation="vertical" />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a navbar with a collapse', () => {
    const {container} = render(
      <Navbar>
        <Navbar.Brand>Brand</Navbar.Brand>
        <Navbar.Collapse id="collapse">
          <Navbar.Nav>
            <Navbar.Item>A</Navbar.Item>
            <Navbar.Item>B</Navbar.Item>
            <Navbar.Item>C</Navbar.Item>
          </Navbar.Nav>
        </Navbar.Collapse>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a navbar with a toggle button', () => {
    const {container} = render(
      <Navbar>
        <Navbar.Brand>Brand</Navbar.Brand>
        <Navbar.Toggle targetId="navCollapse">
          <icons.Hamburger />
        </Navbar.Toggle>
        <Navbar.Collapse id="collapse">
          <Navbar.Nav>
            <Navbar.Item>A</Navbar.Item>
            <Navbar.Item>B</Navbar.Item>
            <Navbar.Item>C</Navbar.Item>
          </Navbar.Nav>
        </Navbar.Collapse>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });
});