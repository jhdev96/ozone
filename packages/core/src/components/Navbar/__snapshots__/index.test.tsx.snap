// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> adds space between the brand and the links 1`] = `
<div>
  <nav
    class="nav horizontal space-between light lg"
  >
    <a
      class="navBrand"
    >
      Brand
    </a>
    <ul
      class="navContent"
    >
      <li
        class="navItem"
      >
        Item 1
      </li>
      <li
        class="navItem"
      >
        Item 2
      </li>
      <li
        class="navItem"
      >
        Item 3
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<Navbar /> centers the navbar items 1`] = `
<div>
  <nav
    class="nav horizontal center light lg"
  >
    <a
      class="navBrand"
    >
      Brand
    </a>
  </nav>
</div>
`;

exports[`<Navbar /> places the navbar items at the end 1`] = `
<div>
  <nav
    class="nav horizontal end light lg"
  >
    <a
      class="navBrand"
    >
      Brand
    </a>
  </nav>
</div>
`;

exports[`<Navbar /> renders a navbar with a brand link 1`] = `
<div>
  <nav
    class="nav horizontal start light lg"
  >
    <a
      class="navBrand"
    >
      Brand
    </a>
  </nav>
</div>
`;

exports[`<Navbar /> renders a navbar with a collapse 1`] = `
<div>
  <nav
    class="nav horizontal start light lg"
  >
    <a
      class="navBrand"
    >
      Brand
    </a>
    <div
      class="navCollapse"
      id="collapse"
    >
      <ul
        class="navContent"
      >
        <li
          class="navItem"
        >
          A
        </li>
        <li
          class="navItem"
        >
          B
        </li>
        <li
          class="navItem"
        >
          C
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`<Navbar /> renders a navbar with a toggle button 1`] = `
<div>
  <nav
    class="nav horizontal start light lg"
  >
    <a
      class="navBrand"
    >
      Brand
    </a>
    <button
      class="navToggle"
      type="button"
    >
      <svg
        height="10"
        style="align-self: center;"
        viewBox="0 0 16 9.5"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          data-name="Group 4"
          id="Group_4"
          transform="translate(-452 -68)"
        >
          <rect
            height="1.5"
            id="Rectangle_703"
            rx="0.75"
            transform="translate(452 68)"
            width="16"
          />
          <rect
            height="1.5"
            id="Rectangle_706"
            rx="0.75"
            transform="translate(452 72)"
            width="16"
          />
          <rect
            height="1.5"
            id="Rectangle_707"
            rx="0.75"
            transform="translate(452 76)"
            width="16"
          />
        </g>
      </svg>
    </button>
    <div
      class="navCollapse"
      id="collapse"
    >
      <ul
        class="navContent"
      >
        <li
          class="navItem"
        >
          A
        </li>
        <li
          class="navItem"
        >
          B
        </li>
        <li
          class="navItem"
        >
          C
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`<Navbar /> renders a navbar with horizontal orientation by default 1`] = `
<div>
  <nav
    class="nav horizontal start light lg"
    data-testid="test"
  >
    <ul
      class="navContent"
    >
      <li
        class="navItem"
      >
        <a
          class="navLink"
          href="#"
        >
          Link 1
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<Navbar /> sets the \`justify\` prop to \`start\` by default 1`] = `
<div>
  <nav
    class="nav horizontal start light lg"
    data-testid="test"
  />
</div>
`;

exports[`<Navbar /> sets the \`orientation\` prop to \`vertical\` 1`] = `
<div>
  <nav
    class="nav vertical start light lg"
  />
</div>
`;
